From d248da4024fbf9b5996c61ffa0af07436b4ec634 Mon Sep 17 00:00:00 2001
From: kk <kk@ubuntu>
Date: Thu, 7 May 2026 10:41:55 +0800
Subject: [PATCH] drm/rockchip: add vop/vop2 bcsh crtc controls

Expose contrast, saturation, and hue controls on RK356x VOP2 CRTCs and program the VP BCSH block from atomic state.
---
 drivers/gpu/drm/rockchip/rockchip_drm_drv.h  |   4 +
 drivers/gpu/drm/rockchip/rockchip_drm_vop.c  | 186 ++++++++++++++++
 drivers/gpu/drm/rockchip/rockchip_drm_vop.h  |  13 ++
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c | 215 +++++++++++++++++++
 drivers/gpu/drm/rockchip/rockchip_vop_reg.c  |  26 +++
 5 files changed, 444 insertions(+)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_drv.h b/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
index 8d566fcd80a2..d68983cd1f95 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_drv.h
@@ -53,6 +53,10 @@ struct rockchip_crtc_state {
 	u32 bus_format;
 	u32 bus_flags;
 	int color_space;
+	int brightness;
+	int contrast;
+	int saturation;
+	int hue;
 };
 #define to_rockchip_crtc_state(s) \
 		container_of(s, struct rockchip_crtc_state, base)
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
index 69900138295b..6d746016aa1e 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.c
@@ -7,6 +7,7 @@
 #include <linux/clk.h>
 #include <linux/component.h>
 #include <linux/delay.h>
 #include <linux/iopoll.h>
 #include <linux/kernel.h>
 #include <linux/log2.h>
@@ -28,6 +29,7 @@
 #include <drm/drm_gem_atomic_helper.h>
 #include <drm/drm_gem_framebuffer_helper.h>
 #include <drm/drm_probe_helper.h>
 #include <drm/drm_self_refresh_helper.h>
 #include <drm/drm_vblank.h>
 
@@ -106,6 +108,17 @@
 #define to_vop(x) container_of(x, struct vop, crtc)
 #define to_vop_win(x) container_of(x, struct vop_win, base)
 
+#define VOP_BCSH_PROP_MIN		0
+#define VOP_BCSH_PROP_DEFAULT		50
+#define VOP_BCSH_PROP_MAX		100
+
+enum vop_bcsh_out_mode {
+	VOP_BCSH_OUT_MODE_BLACK,
+	VOP_BCSH_OUT_MODE_BLUE,
+	VOP_BCSH_OUT_MODE_COLOR_BAR,
+	VOP_BCSH_OUT_MODE_NORMAL_VIDEO,
+};
+
 #define AFBC_FMT_RGB565		0x0
 #define AFBC_FMT_U8U8U8U8	0x5
 #define AFBC_FMT_U8U8U8		0x4
@@ -184,6 +197,11 @@ struct vop {
 	/* optional internal rgb encoder */
 	struct rockchip_rgb *rgb;
 
+	struct drm_property *brightness_prop;
+	struct drm_property *contrast_prop;
+	struct drm_property *saturation_prop;
+	struct drm_property *hue_prop;
+
 	struct vop_win win[];
 };
 
@@ -228,6 +246,72 @@ static void vop_reg_set(struct vop *vop, const struct vop_reg *reg,
 		writel(v, vop->regs + offset);
 }
 
+static bool vop_has_bcsh(struct vop *vop)
+{
+	return VOP_HAS_REG(vop, common, bcsh_en);
+}
+
+static void vop_bcsh_setup(struct drm_crtc *crtc)
+{
+	struct rockchip_crtc_state *s = to_rockchip_crtc_state(crtc->state);
+	struct vop *vop = to_vop(crtc);
+	const struct vop_common *common = vop->data->common;
+	int brightness, contrast, saturation, hue, sin_hue, cos_hue;
+	bool bcsh_en;
+
+	if (!vop_has_bcsh(vop))
+		return;
+
+	bcsh_en = s->brightness != VOP_BCSH_PROP_DEFAULT ||
+		  s->contrast != VOP_BCSH_PROP_DEFAULT ||
+		  s->saturation != VOP_BCSH_PROP_DEFAULT ||
+		  s->hue != VOP_BCSH_PROP_DEFAULT;
+
+	if (!bcsh_en) {
+		VOP_REG_SET(vop, common, bcsh_en, 0);
+		return;
+	}
+
+	if (vop->data->feature & VOP_FEATURE_OUTPUT_RGB10)
+		brightness = fixp_linear_interpolate(0, -128, 100, 127,
+						     s->brightness);
+	else if (vop->data->version == VOP_VERSION(2, 6))
+		brightness = fixp_linear_interpolate(0, -64, 100, 63,
+						     s->brightness);
+	else
+		brightness = fixp_linear_interpolate(0, -32, 100, 31,
+						     s->brightness);
+
+	if (common->bcsh_contrast.mask == 0x1ff) {
+		contrast = fixp_linear_interpolate(0, 0, 100, 511, s->contrast);
+		saturation = fixp_linear_interpolate(0, 0, 100, 511, s->saturation);
+		hue = fixp_linear_interpolate(0, -30, 100, 30, s->hue);
+		sin_hue = fixp_sin32(hue) >> 23;
+		cos_hue = fixp_cos32(hue) >> 23;
+		VOP_REG_SET(vop, common, bcsh_sat_con,
+			    saturation * contrast / 0x100);
+	} else {
+		contrast = fixp_linear_interpolate(0, 0, 100, 255, s->contrast);
+		saturation = fixp_linear_interpolate(0, 0, 100, 255, s->saturation);
+		hue = fixp_linear_interpolate(0, -30, 100, 30, s->hue);
+		sin_hue = fixp_sin32(hue) >> 24;
+		cos_hue = fixp_cos32(hue) >> 24;
+		VOP_REG_SET(vop, common, bcsh_sat_con,
+			    saturation * contrast / 0x80);
+	}
+
+	VOP_REG_SET(vop, common, bcsh_r2y_en, 1);
+	VOP_REG_SET(vop, common, bcsh_y2r_en, 1);
+	VOP_REG_SET(vop, common, bcsh_r2y_csc_mode, 0);
+	VOP_REG_SET(vop, common, bcsh_y2r_csc_mode, 0);
+	VOP_REG_SET(vop, common, bcsh_brightness, brightness);
+	VOP_REG_SET(vop, common, bcsh_contrast, contrast);
+	VOP_REG_SET(vop, common, bcsh_sin_hue, sin_hue);
+	VOP_REG_SET(vop, common, bcsh_cos_hue, cos_hue);
+	VOP_REG_SET(vop, common, bcsh_out_mode, VOP_BCSH_OUT_MODE_NORMAL_VIDEO);
+	VOP_REG_SET(vop, common, bcsh_en, 1);
+}
+
 static inline uint32_t vop_get_intr_type(struct vop *vop,
 					 const struct vop_reg *reg, int type)
 {
@@ -1459,6 +1543,7 @@ static void vop_crtc_atomic_enable(struct drm_crtc *crtc,
 	}
 
 	VOP_REG_SET(vop, common, out_mode, s->output_mode);
+	vop_bcsh_setup(crtc);
 
 	VOP_REG_SET(vop, modeset, htotal_pw, (htotal << 16) | hsync_len);
 	val = hact_st << 16;
@@ -1581,6 +1666,7 @@ static void vop_crtc_atomic_flush(struct drm_crtc *crtc,
 	/* Enable AFBC if there is some AFBC window, disable otherwise. */
 	s = to_rockchip_crtc_state(crtc->state);
 	VOP_AFBC_SET(vop, enable, s->enable_afbc);
+	vop_bcsh_setup(crtc);
 	vop_cfg_done(vop);
 
 	/* Ack the DMA transfer of the previous frame (RK3066). */
@@ -1668,6 +1754,13 @@ static void vop_crtc_reset(struct drm_crtc *crtc)
 		__drm_atomic_helper_crtc_reset(crtc, &crtc_state->base);
 	else
 		__drm_atomic_helper_crtc_reset(crtc, NULL);
+
+	if (crtc_state) {
+		crtc_state->brightness = VOP_BCSH_PROP_DEFAULT;
+		crtc_state->contrast = VOP_BCSH_PROP_DEFAULT;
+		crtc_state->saturation = VOP_BCSH_PROP_DEFAULT;
+		crtc_state->hue = VOP_BCSH_PROP_DEFAULT;
+	}
 }
 
 #ifdef CONFIG_DRM_ANALOGIX_DP
@@ -1735,11 +1828,60 @@ vop_crtc_verify_crc_source(struct drm_crtc *crtc, const char *source_name,
 }
 #endif
 
+static int vop_crtc_atomic_set_property(struct drm_crtc *crtc,
+					struct drm_crtc_state *state,
+					struct drm_property *property,
+					uint64_t val)
+{
+	struct vop *vop = to_vop(crtc);
+	struct rockchip_crtc_state *s = to_rockchip_crtc_state(state);
+
+	if (val > VOP_BCSH_PROP_MAX)
+		return -EINVAL;
+
+	if (property == vop->brightness_prop)
+		s->brightness = val;
+	else if (property == vop->contrast_prop)
+		s->contrast = val;
+	else if (property == vop->saturation_prop)
+		s->saturation = val;
+	else if (property == vop->hue_prop)
+		s->hue = val;
+	else
+		return -EINVAL;
+
+	return 0;
+}
+
+static int vop_crtc_atomic_get_property(struct drm_crtc *crtc,
+					const struct drm_crtc_state *state,
+					struct drm_property *property,
+					uint64_t *val)
+{
+	struct vop *vop = to_vop(crtc);
+	struct rockchip_crtc_state *s = to_rockchip_crtc_state(state);
+
+	if (property == vop->brightness_prop)
+		*val = s->brightness;
+	else if (property == vop->contrast_prop)
+		*val = s->contrast;
+	else if (property == vop->saturation_prop)
+		*val = s->saturation;
+	else if (property == vop->hue_prop)
+		*val = s->hue;
+	else
+		return -EINVAL;
+
+	return 0;
+}
+
 static const struct drm_crtc_funcs vop_crtc_funcs = {
 	.set_config = drm_atomic_helper_set_config,
 	.page_flip = drm_atomic_helper_page_flip,
 	.destroy = drm_crtc_cleanup,
 	.reset = vop_crtc_reset,
+	.atomic_set_property = vop_crtc_atomic_set_property,
+	.atomic_get_property = vop_crtc_atomic_get_property,
 	.atomic_duplicate_state = vop_crtc_duplicate_state,
 	.atomic_destroy_state = vop_crtc_destroy_state,
 	.enable_vblank = vop_crtc_enable_vblank,
@@ -1853,6 +1995,35 @@ static void vop_plane_add_properties(struct drm_plane *plane,
 						   DRM_MODE_ROTATE_0 | flags);
 }
 
+static int vop_create_bcsh_properties(struct vop *vop)
+{
+	if (!vop_has_bcsh(vop))
+		return 0;
+
+	vop->brightness_prop = drm_property_create_range(vop->drm_dev, 0,
+							 "brightness",
+							 VOP_BCSH_PROP_MIN,
+							 VOP_BCSH_PROP_MAX);
+	vop->contrast_prop = drm_property_create_range(vop->drm_dev, 0,
+						       "contrast",
+						       VOP_BCSH_PROP_MIN,
+						       VOP_BCSH_PROP_MAX);
+	vop->saturation_prop = drm_property_create_range(vop->drm_dev, 0,
+							 "saturation",
+							 VOP_BCSH_PROP_MIN,
+							 VOP_BCSH_PROP_MAX);
+	vop->hue_prop = drm_property_create_range(vop->drm_dev, 0,
+						  "hue",
+						  VOP_BCSH_PROP_MIN,
+						  VOP_BCSH_PROP_MAX);
+
+	if (!vop->brightness_prop || !vop->contrast_prop ||
+	    !vop->saturation_prop || !vop->hue_prop)
+		return -ENOMEM;
+
+	return 0;
+}
+
 static int vop_create_crtc(struct vop *vop)
 {
 	const struct vop_data *vop_data = vop->data;
@@ -1904,6 +2075,21 @@ static int vop_create_crtc(struct vop *vop)
 		goto err_cleanup_planes;
 
 	drm_crtc_helper_add(crtc, &vop_crtc_helper_funcs);
+	ret = vop_create_bcsh_properties(vop);
+	if (ret)
+		goto err_cleanup_crtc;
+	if (vop->brightness_prop)
+		drm_object_attach_property(&crtc->base, vop->brightness_prop,
+					   VOP_BCSH_PROP_DEFAULT);
+	if (vop->contrast_prop)
+		drm_object_attach_property(&crtc->base, vop->contrast_prop,
+					   VOP_BCSH_PROP_DEFAULT);
+	if (vop->saturation_prop)
+		drm_object_attach_property(&crtc->base, vop->saturation_prop,
+					   VOP_BCSH_PROP_DEFAULT);
+	if (vop->hue_prop)
+		drm_object_attach_property(&crtc->base, vop->hue_prop,
+					   VOP_BCSH_PROP_DEFAULT);
 	if (vop->lut_regs) {
 		drm_mode_crtc_set_gamma_size(crtc, vop_data->lut_size);
 		drm_crtc_enable_color_mgmt(crtc, 0, false, vop_data->lut_size);
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.h b/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
index 0cf512cc1614..aabe65835650 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
@@ -125,6 +125,19 @@ struct vop_common {
 	struct vop_reg dma_stop;
 	struct vop_reg out_mode;
 	struct vop_reg standby;
+
+	struct vop_reg bcsh_brightness;
+	struct vop_reg bcsh_contrast;
+	struct vop_reg bcsh_sat_con;
+	struct vop_reg bcsh_sin_hue;
+	struct vop_reg bcsh_cos_hue;
+	struct vop_reg bcsh_r2y_csc_mode;
+	struct vop_reg bcsh_r2y_en;
+	struct vop_reg bcsh_y2r_csc_mode;
+	struct vop_reg bcsh_y2r_en;
+	struct vop_reg bcsh_color_bar;
+	struct vop_reg bcsh_out_mode;
+	struct vop_reg bcsh_en;
 };
 
 struct vop_misc {
index 9873172e3fd3..21e96469e6fc 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop_reg.c
+++ b/drivers/gpu/drm/rockchip/rockchip_vop_reg.c
@@ -275,6 +275,19 @@ static const struct vop_common px30_common = {
 	.dither_down_sel = VOP_REG(PX30_DSP_CTRL2, 0x1, 7),
 	.dither_down_mode = VOP_REG(PX30_DSP_CTRL2, 0x1, 6),
 	.cfg_done = VOP_REG_SYNC(PX30_REG_CFG_DONE, 0x1, 0),
+
+	.bcsh_en = VOP_REG(PX30_BCSH_CTRL, 0x1, 0),
+	.bcsh_r2y_csc_mode = VOP_REG(PX30_BCSH_CTRL, 0x1, 1),
+	.bcsh_out_mode = VOP_REG(PX30_BCSH_CTRL, 0x3, 2),
+	.bcsh_y2r_csc_mode = VOP_REG(PX30_BCSH_CTRL, 0x3, 4),
+	.bcsh_y2r_en = VOP_REG(PX30_BCSH_CTRL, 0x1, 6),
+	.bcsh_r2y_en = VOP_REG(PX30_BCSH_CTRL, 0x1, 7),
+	.bcsh_color_bar = VOP_REG(PX30_BCSH_COL_BAR, 0xffffff, 0),
+	.bcsh_brightness = VOP_REG(PX30_BCSH_BCS, 0xff, 0),
+	.bcsh_contrast = VOP_REG(PX30_BCSH_BCS, 0x1ff, 8),
+	.bcsh_sat_con = VOP_REG(PX30_BCSH_BCS, 0x3ff, 20),
+	.bcsh_sin_hue = VOP_REG(PX30_BCSH_H, 0x1ff, 0),
+	.bcsh_cos_hue = VOP_REG(PX30_BCSH_H, 0x1ff, 16),
 };
 
 static const struct vop_modeset px30_modeset = {
@@ -1145,6 +1158,19 @@ static const struct vop_common rk3328_common = {
 	.dsp_blank = VOP_REG(RK3328_DSP_CTRL0, 0x3, 18),
 	.out_mode = VOP_REG(RK3328_DSP_CTRL0, 0xf, 0),
 	.cfg_done = VOP_REG_SYNC(RK3328_REG_CFG_DONE, 0x1, 0),
+
+	.bcsh_brightness = VOP_REG(RK3328_BCSH_BCS, 0xff, 0),
+	.bcsh_contrast = VOP_REG(RK3328_BCSH_BCS, 0x1ff, 8),
+	.bcsh_sat_con = VOP_REG(RK3328_BCSH_BCS, 0x3ff, 20),
+	.bcsh_out_mode = VOP_REG(RK3328_BCSH_BCS, 0x3, 30),
+	.bcsh_sin_hue = VOP_REG(RK3328_BCSH_H, 0x1ff, 0),
+	.bcsh_cos_hue = VOP_REG(RK3328_BCSH_H, 0x1ff, 16),
+	.bcsh_r2y_csc_mode = VOP_REG(RK3328_BCSH_CTRL, 0x3, 6),
+	.bcsh_r2y_en = VOP_REG(RK3328_BCSH_CTRL, 0x1, 4),
+	.bcsh_y2r_csc_mode = VOP_REG(RK3328_BCSH_CTRL, 0x3, 2),
+	.bcsh_y2r_en = VOP_REG(RK3328_BCSH_CTRL, 0x1, 0),
+	.bcsh_color_bar = VOP_REG(RK3328_BCSH_COLOR_BAR, 0xffffff, 8),
+	.bcsh_en = VOP_REG(RK3328_BCSH_COLOR_BAR, 0x1, 0),
 };
 
 static const struct vop_intr rk3328_vop_intr = {
-- 
2.34.1

